File: scala/src/main/scala/banking/VerifiedArithmetic.scala
package banking

object VerifiedArithmetic:
  def calculateNewBalance(balance: Int, amount: Int): Int = {
    require(balance >= 0)
    require(amount >= 0)
    require(amount <= balance)
    balance - amount
  }.ensuring(result => result >= 0 && result == balance - amount)

File: python/src/banking/verified_arithmetic.py
"""Small pure functions suitable for symbolic analysis."""


def calculate_new_balance(balance: int, amount: int) -> int:
    """Calculate a valid post-withdrawal balance.

    pre: balance >= 0
    pre: amount >= 0
    pre: amount <= balance
    post: __return__ >= 0
    post: __return__ == balance - amount
    """
    return balance - amount
